lib/IRGen/GenProto.cpp

```diff
@@ -2114,6 +2114,46 @@ static bool isConstantWitnessTable(SILWitnessTable *wt) {
   return true;
 }
 
+static void noteUseOfTypeMetadataInConditionalRequirements(
```

Comment:
Why isn't this handled by emitting those mangled names in the witness table?

Reply:
This is the witness table:

```
private enum E {
}

private protocol P {
  associatedtype AT
}

private struct S<A, B> {
  init()
}

extension S : P where A == E {
  typealias AT = B
}
sil_witness_table private <A, B where A == E> S<A, B>: P module ProtoCrash {
  associated_type AT: B
}
```

Which mangled names are you referring to?

Reply:
Here is the ordering of code:
```
    // Okay, emit any definitions that we suddenly need.                        
    irgen.emitLazyDefinitions();                                                
                                                                                
    // Register our info with the runtime if needed.                            
    if (Opts.UseJIT) {                                                          
      IGM.emitBuiltinReflectionMetadata();                                      
      IGM.emitRuntimeRegistration();                                            
    } else {                                                                    
      // Emit protocol conformances into a section we can recognize at runtime. 
      // In JIT mode these are manually registered above.                       
      IGM.emitSwiftProtocols();                                                 
      IGM.emitProtocolConformances();
```

IGM.emitProtocolConformances will create typedefs (which note the use of type metadata) but this is too late.

Reply:
Does it work if you emit the lazy definitions after protocol conformances?

We might need a loop to pump this until it stops adding new lazy definitions. I'm thinking of a case where emitting conditional requirements triggers lazy metadata emission, which triggers emitting another witness table, and so on.